extensions/src/main/java/org/pentaho/platform/plugin/action/ActionInvokerSystemListener.java

```diff
@@ -0,0 +1,156 @@
+          if ( jsonObject != null ) {
+            dotJsons.put( jsonFile.getAbsolutePath(), jsonObject );
+          }
+        }
+      }
+    }
+    return dotJsons;
+  }
+
+  private JSONObject getJsonFromFile( final File file ) throws IOException {
+    FileReader fr = new FileReader( file );
+    try {
+      JSONParser jsonParser = new JSONParser();
+      JSONObject jsonObject = (JSONObject) jsonParser.parse( fr );
+      String jsonString = jsonObject.toJSONString();
+      fr.close();
+      return (JSONObject) jsonParser.parse( jsonString );
+    } catch ( ParseException e ) {
+      logger.error( "File: " + file.getAbsolutePath() + " is not valid json, and will not be processed" );
+      fr.close();
+      //rename the file here and allow the chronos container to die because of bad input
+      renameFile( file.getAbsolutePath() );
+    }
+    return null;
+  }
+
+  private File[] listFiles( final File folder, final String fileExtension ) throws IOException {
+    if ( folder.isDirectory() && folder.canRead() ) {
+      return folder.listFiles( new FileFilter() {
+        @Override
+        public boolean accept( File f ) {
+          return f.isFile() && f.getName().toLowerCase().endsWith( fileExtension );
+        }
+      } );
+    } else {
+      logger.error( folder.getAbsolutePath() + " is not a valid directory" );
+    }
+    return null;
+  }
+
+  private void renameFile( final String filename ) {
+    File dotJson = new File( filename );
+    File dotJsonDotCurTime = new File( filename + "." + System.currentTimeMillis() );
+    dotJson.renameTo( dotJsonDotCurTime );
+  }
+
+  /**
+   * System listener looks for JSON files at start up and runs them as work items
+   * @param session
+   * @return
+   */
+  @Override
+  public boolean startup( IPentahoSession session ) {
+    Response status;
+    Map<String, JSONObject> jsonFiles = new HashMap<>(); //map between the json objects and the filenames from which they were consumed
+    try {
+      if ( environmentVariablesFilepath != null ) {
+        JSONObject environmentVariables = getJsonFromFile( new File( environmentVariablesFilepath ) );
+        jsonFiles.put( environmentVariablesFilepath, environmentVariables );
+      } else {
+        jsonFiles = parseJsonFiles();
+      }
+      for ( String filename : jsonFiles.keySet() ) {
+        JSONObject json = jsonFiles.get( filename );
+        WorkItemRequestPayload payload = new WorkItemRequestPayload( json );
+        if ( !payload.isValid() ) {
+          logger.error( "File: " + filename + " is missing required parameters, skipping" );
+        } else {
+          status = issueRequest( payload );
+          if ( status == null ) {
+            logger.error( "File: " + filename + " has parameters formatted in an unparsable way. Could not process work item." );
+          }
+          //todo- sometype of status handling once WI lifecycle is defined
+        }
+        renameFile( filename );
```

Comment:
Suggestion: adding a try-finally clause inside the for loop, so that whatever ends up happening ( successful execution, NPE or some other unchecked exception ), we'll always be renaming the file, appending it the timestamp:

`for ( String filename : jsonFiles.keySet() ) {     
     
     try {
         JSONObject json = jsonFiles.get( filename );
         WorkItemRequestPayload payload = new WorkItemRequestPayload( json );
         if ( !payload.isValid() ) {
           logger.error( "File: " + filename + " is missing required parameters, skipping" );
         } else {
           status = issueRequest( payload );
           if ( status == null ) {
             logger.error( "File: " + filename + " has parameters formatted in an unparsable way. Could not process work item." );
           }
           //todo- sometype of status handling once WI lifecycle is defined
         }
      } finally {
         renameFile( filename );
      }
}`